### PR TITLE
Clarify behavior when RunServerCompilation throws an exception

### LINKED ISSUE #7866: Figure out what should be done in response to exceptions from RunServerCompilation and HandleResponse in BuildClient
See https://github.com/dotnet/roslyn/pull/7841

### PR DESCRIPTION
**Pull request by jaredpar**
A question was raised about the behavior of the compiler when
RunServerCompilation throws an exception.  After looking through the
code I determined the best answer is to let the exception propogate (or
possibly promptly terminate the process).  The API is designed to never
throw, exception in cases of cancellation.  Errors communicating with
the server should instead be handled in the return value directly.

closes #7866

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler

File: src/Compilers/Core/CommandLine/BuildClient.cs [hunk 1]
```diff
@@ -58,6 +59,30 @@ internal RunCompilationResult(int exitCode, bool ranOnServer = false)
         }
     }
 
+    internal struct RunServerResult
+    {
+        internal static readonly RunServerResult Failed = new RunServerResult();
+
+        private BuildResponse _buildResponse;
```
Comment (AlekseyTs) [inline src/Compilers/Core/CommandLine/BuildClient.cs]: readonly?

File: src/Compilers/Core/CommandLine/BuildClient.cs [hunk 2]
```diff
@@ -149,7 +166,11 @@ public Task<RunCompilationResult> RunCompilationAsync(IEnumerable<string> origin
 
         protected abstract int RunLocalCompilation(string[] arguments, BuildPaths buildPaths, TextWriter textWriter);
 
-        protected abstract Task<BuildResponse> RunServerCompilation(List<string> arguments, BuildPaths buildPaths, string sessionName, string keepAlive, string libDirectory, CancellationToken cancellationToken);
+        /// <summary>
+        /// Attempts to run the compilation on the server.  This function should never return a Task in a faulted
+        /// state.  The task should only return in a completed or cancelled state.
```
Comment (AlekseyTs) [inline src/Compilers/Core/CommandLine/BuildClient.cs]: It doesn't look like we rely on any async behavior here, why return a Task at all?

File: src/Compilers/Shared/CoreClrBuildClient.cs [hunk 3]
```diff
@@ -46,14 +46,22 @@ protected override string GetSessionKey(BuildPaths buildPaths)
             return string.Empty;
         }
 
-        protected override async Task<BuildResponse> RunServerCompilation(List<string> arguments, BuildPaths buildPaths, string pipeName, string keepAlive, string libDirectory, CancellationToken cancellationToken)
+        protected override async Task<RunServerResult> RunServerCompilation(List<string> arguments, BuildPaths buildPaths, string pipeName, string keepAlive, string libDirectory, CancellationToken cancellationToken)
```
Comment (AlekseyTs) [inline src/Compilers/Shared/CoreClrBuildClient.cs]: It looks like this implementation actually can return a Task in an incomplete state.

File: src/Compilers/Core/CommandLine/BuildClient.cs (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/Core/CommandLine/BuildClient.cs]: The portable implementation of this method uses `async` today.  

Ideally the desktop version would as well.  Unfortunately though the version of .NET we use in master (4.5.0) doesn't have async methods for Named pipes.  Hence we're limited to sync behavior.  In future branch we plan to use async versions here.

File: src/Compilers/Core/CommandLine/BuildClient.cs (hunk 1 above)
Comment (jaredpar) [inline src/Compilers/Core/CommandLine/BuildClient.cs]: Yep.

File: src/Compilers/Shared/CoreClrBuildClient.cs (hunk 3 above)
Comment (jaredpar) [inline src/Compilers/Shared/CoreClrBuildClient.cs]: How so?  There are two `return` paths here and all return a valid `RunServerResult` object.  That should always result in a completed `Task` value (second is in a blanket `catch` block).

Comment (AlekseyTs) [inline src/Compilers/Shared/CoreClrBuildClient.cs]: There is an await prior to return.

Comment (jaredpar) [inline src/Compilers/Shared/CoreClrBuildClient.cs]: Agree.  But the caller uses `Task.Result` which blocks until the value completes.  

Is the wording of the comment in the other file what's bothering you here?

Comment (AlekseyTs) [inline src/Compilers/Shared/CoreClrBuildClient.cs]: Yes, the claim that the task is complete.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/Shared/CoreClrBuildClient.cs (hunk 3 above)
Comment (jaredpar) [inline src/Compilers/Shared/CoreClrBuildClient.cs]: Gotcha.  I will clean that up.

**Comment by jaredpar** [comment]
Closing and opening a new PR on stabilization (with a much different approach)

**Comment by jaredpar** [comment]
New PR is #8776

**Pull request closed without merging**